src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java

```diff
@@ -820,17 +811,16 @@ public void onQueryPhase(SearchContext searchContext, long tookInNanos) {
                         return;
                     }
 
-                    @SuppressWarnings("unchecked")
-                    final Map<String, Set<String>> maskedFieldsMap = (Map<String, Set<String>>) HeaderHelper.deserializeSafeFromHeader(
-                        threadPool.getThreadContext(),
-                        ConfigConstants.OPENDISTRO_SECURITY_MASKED_FIELD_HEADER
-                    );
-                    final String maskedEval = SecurityUtils.evalMap(maskedFieldsMap, indexModule.getIndex().getName());
-                    if (maskedEval != null) {
-                        final Set<String> mf = maskedFieldsMap.get(maskedEval);
-                        if (mf != null && !mf.isEmpty()) {
+                    try {
+                        if (dlsFlsValve.hasFieldMasking(indexModule.getIndex().getName())) {
                             dlsFlsValve.onQueryPhase(queryResult);
                         }
+                    } catch (PrivilegesEvaluationException e) {
+                        log.error("Error while evaluating field masking config", e);
+                        // It is safe to call the code nevertheless, as this code does not enforce any privileges.
+                        // Rather, it performs some fixes to keep aggregations happy after field masking has been
+                        // applied. If no field masking has been applied, this should be a no-op.
+                        dlsFlsValve.onQueryPhase(queryResult);
```

Comment:
Then it would be called twice for the success case when there's field masking active - first in line 816 and then in the finally block.

The logic is: Call it:

- Called if `hasFieldMasking` returns true
- Called if `hasFieldMasking` throws exception
- NOT called if `hasFieldMasking` returns false

Reply:
I meant move the logic completely to finally block since it is called in catch upon expecting PrivilegesEvaluationException.

Reply:
In the finally block, it would be also executed if `hasFieldMasking` throws now exception, but returns false. This is the 90% case - executing this logic would have slow things down unnecessarily then.